### DISCUSSION TITLE
Simplest way to connect a view to its view model?

### DISCUSSION THREAD

**Discussion by ero-qt** [AvaloniaUI/Avalonia#14848] [Q&A] [answered]
Hi everyone.

I was wondering if this code from the MVVM template in `App.axaml.cs` is really necessary:
```cs
public override void OnFrameworkInitializationCompleted()
{
    if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
    {
        desktop.MainWindow = new MainWindow
        {
            DataContext = new MainWindowViewModel(),
        };
    }

    base.OnFrameworkInitializationCompleted();
}
```

I assumed, since I specify `x:DataType="vm:MainWindowViewModel"` in `MainWindow.axaml`, I wouldn't need to do anything else. Evidently, removing that code from `App.axaml.cs` results in the app not starting.

Why is this? And what's the easiest way to simply connect my views to their view models? I have 2 separate C# projects for them, if that matters.

**Comment by BAndysc** [comment, marked as answer]
The `x:DataType` directive is a hint for the XAML compiler to enable compiled bindings and for providing completion suggestions in the IDE. It doesn't instantiate a viewmodel.

You can instantiate a view model from the axaml like that:
```
<Window ...>
    <Window.DataContext>
        <MyViewModelType />
    </Window.DataContext>
</Window>
```

but this works only if the view model has a default constructor, which is almost never the case. Otherwise you can explicitly assign a ViewModel in the View constructor, but that's not really something you want.

---

Generally in MVVM, there are two main approaches to linking views and view models: "ViewModel First" and "View First". In the "View First" approach, you instantiate a view explicitly (either in the code using `new MyUserControl` or in the axaml) and the viewmodel is expected to be found automatically. In the "ViewModel First" approach it works the other way around - you explicitly create ViewModels and Views are expected to be created automatically. Since the application logic should be in View Models rather than in Views, I think the "view first" is always an inferior option. ViewModel First approach works well with Avalonia using Data Templates, as seen in the default MVVM template with the `ViewLocator.cs`. This IDataTemplate finds a matching view type for a ViewModel and instantiates the view.

However, to initiate this process, a starting point is required, which is what the code you posted achieves by explicitly setting the DataContext. With the ViewLocator, beyond this initial point, manual DataContext assignments are not needed, as view locator automatically connects views and view models based on naming conventions.

Suppose your "root" MainWindowViewModel has a list of `DocumentViewModels`

```
class MainWindowViewModel {
    public ObservableCollection<DocumentViewModel> Documents { get; }
}
```

in the main view you can bind the list
```
<Window>...
    <TabControl ItemsSource="{CompiledBinding Documents}">
    </TabControl>
</Window>
```

and if you have the ViewLocator and a view with name `DocumentView` - the view will be automatically instantiated, which is what you want, right?

**Reply by ero-qt** [reply]
This makes sense to me (although I find it strangely designed).

I am, however, having some trouble. My current architecture is the following:
- `MyProject.App`: contains `Program.cs`, `App.axaml`, `App.axaml.cs`, `ViewLocator.cs`,
- `MyProject.ViewModels`: contains `ViewModelBase.cs`, `MainWindowViewModel.cs`, `DocumentViewModel.cs`,
- `MyProject.Views`: contains `MainWindow.axaml`, `MainWindow.axaml.cs`, `DocumentView.axaml`, `DocumentView.axaml.cs`.

Now, the app runs. However, the `DocumentView`s aren't being rendered.  
Here are the relevant contents:  

<details><summary><code>DocumentView.axaml</code></summary>

```xml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:vm="using:MyProject.ViewModels"

             mc:Ignorable="d"
             x:Class="MyProject.Views.DocumentView"
             x:DataType="vm:DocumentViewModel">

    <TextBlock Text="{Binding Text}" />

</UserControl>
```
</details>

<details><summary><code>DocumentViewModel.cs</code></summary>

```cs
namespace MyProject.ViewModels;

public class DocumentViewModel : ViewModelBase
{
    public DocumentViewModel(string text)
    {
        Text = text;
    }

    public string Text { get; }
}
```
</details>

<details><summary><code>MainWindow.axaml</code></summary>

```xml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:vm="using:MyProject.ViewModels"

        mc:Ignorable="d"
        x:Class="MyProject.Views.MainWindow"
        x:DataType="vm:MainWindowViewModel">

    <TabControl ItemsSource="{CompiledBinding Documents}" />

</Window>
```
</details>

<details><summary><code>MainWindowViewModel.cs</code></summary>

```cs
using System.Collections.ObjectModel;

namespace MyProject.ViewModels;

public class MainWindowViewModel : ViewModelBase
{
    public ObservableCollection<DocumentViewModel> Documents { get; } =
    [
        new("Document1"),
        new("Document2"),
    ];
}
```
</details>

<details><summary><code>ViewLocator.cs</code></summary>

```cs
using System;

using Avalonia.Controls;
using Avalonia.Controls.Templates;

using MyProject.ViewModels;

namespace MyProject.Templates;

internal sealed class ViewLocator : IDataTemplate
{
    public Control? Build(object? data)
    {
        if (data?.GetType().FullName is not string viewModelName)
        {
            return null;
        }

        string viewName = viewModelName.Replace("ViewModel", "View", StringComparison.Ordinal);
        if (Type.GetType(viewName) is not Type viewType)
        {
            return null;
        }

        if (Activator.CreateInstance(viewType) is not Control view)
        {
            return null;
        }

        view.DataContext = data;
        return view;
    }

    public bool Match(object? data)
    {
        return data is ViewModelBase;
    }
}
```
</details>

Through debugging, I can see that `viewName` in `ViewLocator.Build` is `MyProject.Views.DocumentView`, which is the accurate type name, including the namespace. However, `Type.GetType(viewName)` returns `null`. I assume this is because `Type.GetType(String)` searches the type within its own assembly. `MyProject.Views.DocumentView` is, of course, in the `MyProject.Views` assembly.

What can I do to solve this?

**Reply by timunie** [reply]
have you seen our samples explaining some basics? https://github.com/AvaloniaUI/Avalonia.Samples

**Reply by ero-qt** [reply]
I don't believe any of the samples have this same architecture.

**Reply by BAndysc** [reply]
In order to lookup in all assemblies, you need to provide a [full qualified name](https://learn.microsoft.com/en-us/dotnet/api/system.type.assemblyqualifiedname?view=net-8.0#system-type-assemblyqualifiedname). Change `FullName` to `AssemblyQualifiedName` and it should work

**Reply by ero-qt** [reply]
Thank you, this works perfectly.

A little beyond the topic of this discussion; the tab controls now have the value of the `Document`'s `Text` property as both the title and the content. How would I control the title and contents separately?

**Reply by BAndysc** [reply]
Set its ItemTemplate
```
<TabControl.ItemTemplate>
    <DataTemplate>
        i.e.  <TextBlock Text="{Binding Title}" />
    </DataTemplate>
</TabControl.ItemTemplate>
```

The actual "content" content template can be also set via ~`SelectedContentTemplate`~ `ContentTemplate`. If you don't specify a custom ~`SelectedContentTemplate`~ `ContentTemplate`, then it falls back to parent data templates, eventually matching the ViewLocator DataTemplate.

**Reply by thevortexcloud** [reply]
> SelectedContentTemplate

You can't actually set it. It's setter is internal. Despite the fact the comment says you can set it.

https://github.com/AvaloniaUI/Avalonia/blob/11f5203ab24515a4094d7140d0d757d5016e2007/src/Avalonia.Controls/TabControl.cs#L130-L140

You wont get a compile warning either if you try to use it anyway. You will however get a runtime error. It confused the hell out of me too when I tried to use awhile ago since the error it gives is not particularly obvious either. I however did not look too far into what it's actually meant to be used for (I assume by Avalonia itself to show which template has been selected?).

**Reply by BAndysc** [reply]
Oh you are right, `SelectedContentTemplate` is just an implementation detail, `ContentTemplate` is the data template for the selected tab (and ItemTemplate for the tab "header"), thanks for correction
